Share connections between replications

sagelywizard opened the pull request:
Prior to this PR, replication pools would unnecessarily hold onto idle connections. The connection pool could also block for a long period of time when terminating. This PR moves connection ownership from the connection pool to a separate process which shares connections between pools. Connections live longer than they are being used, but idle connections are periodically closed on a configurable interval.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,105 @@
+
+
+acquire(URL) ->
+    case gen_server:call(?MODULE, {acquire, URL}) of
```
is 5s long enough? We might need to create a connection, after all.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,105 @@
+            case ets:match_object(?MODULE, #connection{host=Host, port=Port, owner=undefined, _='_'}, 1) of
+                '$end_of_table' ->
+                    {ok, WorkerPid0} = ibrowse:spawn_link_worker_process(URL),
+                    couch_log:error("Worker created! ~p", [WorkerPid0]),
```
notice not error (throughout)

sagelywizard replied:
Blah, I forgot to remove a lot of debugging logging. I'll make some changes.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,105 @@
+        #url{host=Host, port=Port} ->
+            case ets:match_object(?MODULE, #connection{host=Host, port=Port, owner=undefined, _='_'}, 1) of
+                '$end_of_table' ->
+                    {ok, WorkerPid0} = ibrowse:spawn_link_worker_process(URL),
```
we might want this to be async, could take a while to establish a connection, and we're then not responding to other calls.

sagelywizard replied:
I suspect that's the case. I'll do that.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,105 @@
+
+% owner crashed
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
+    Workers = ets:match_object(?MODULE, #connection{owner=Pid, _='_'}),
```
nice!

sagelywizard replied:
🏇

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,105 @@
+handle_info({'EXIT', Pid, Reason}, State) ->
+    case ets:match_object(?MODULE, #connection{worker=Pid, _='_'}, 1) of
+        '$end_of_table' ->
+            ok;
```
do we expect this or is this defensive programming?

sagelywizard replied:
If the owner crashes (which is linked to the worker process) then the workers will be removed from the ets table before the worker crash messages are received.

rnewson commented on src/couch_replicator_httpc_pool.erl:
```diff
@@ -147,9 +132,8 @@ code_change(_OldVsn, #state{}=State, _Extra) ->
-    lists:foreach(fun couch_replicator_connection:relinquish/1, State#state.free),
-    lists:foreach(fun couch_replicator_connection:relinquish/1, State#state.busy).
+terminate(_Reason, _State) ->
+    ok.
```
don't you need to relinquish State#state.conns?

sagelywizard replied:
No, because the connection gen_server monitors the client pool process and relinquishes the connections when it sees the `'DOWN'` message.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -8,9 +8,13 @@
+}).
 
--record(connection, {
+-record(conn, {
```
why rename the record?

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -23,9 +27,11 @@ start_link() ->
 init([]) ->
-    ?MODULE = ets:new(?MODULE, [named_table, public, {keypos, #connection.worker}]),
+    ?MODULE = ets:new(?MODULE, [named_table, public, {keypos, #conn.worker}]),
+    Interval = 60000,
```
should be configurable.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -46,16 +52,16 @@ handle_call({acquire, URL}, From, State) ->
     WorkerPid = case ibrowse_lib:parse_url(URL) of
         #url{host=Host, port=Port} ->
-            case ets:match_object(?MODULE, #connection{host=Host, port=Port, owner=undefined, _='_'}, 1) of
+            case ets:match_object(?MODULE, #conn{host=Host, port=Port, owner=undefined, _='_'}, 1) of
```
renaming this record (to a worse name imo) is causing non-functional changes.

sagelywizard replied:
I can change it back, I don't have a strong opinion.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -46,16 +52,16 @@ handle_call({acquire, URL}, From, State) ->
                     {ok, WorkerPid0} = ibrowse:spawn_link_worker_process(URL),
                     couch_log:error("Worker created! ~p", [WorkerPid0]),
-                    ets:insert(?MODULE, #connection{host=Host, port=Port, owner=Pid, worker=WorkerPid0}),
+                    true = ets:insert_new(?MODULE, #conn{host=Host, port=Port, owner=Pid, worker=WorkerPid0}),
```
yet more changes that aren't related to 'Periodically close idle connections' intent.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -40,10 +45,11 @@ start_link() ->
 init([]) ->
     ?MODULE = ets:new(?MODULE, [named_table, public, {keypos, #conn.worker}]),
-    Interval = 60000,
+    ok = config:listen_for_changes(?MODULE, self()),
```
touché

rnewson commented:
I really like this, great job. That said, lots of comments to address before merging. Very much appreciate the multiple commits, it made reviewing much easier.

sagelywizard commented:
@rnewson I believe I made all the changes from your comments. I'm currently testing, adding more logging, and adding instrumentation.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+-export([handle_config_change/5, handle_config_terminate/3]).
+
+-define(DEFAULT_CLOSE_INTERVAL, 60000).
+-define(DEFAULT_CLOSE_BATCH_SIZE, 50).
```
Do we use this?

sagelywizard replied:
Ah, no, forgot to remove that. Good catch.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+-record(connection, {
+    host,
+    port,
+    worker,
```
Minor nit, is it idiomatically better to put the key as first field in the record?

sagelywizard replied:
Sure, changed.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+    ok = config:listen_for_changes(?MODULE, self()),
+    Interval = config:get_integer("replicator", "connection_close_interval", ?DEFAULT_CLOSE_INTERVAL),
+    {ok, Timer} = timer:send_after(Interval, close_idle_connections),
+    process_flag(trap_exit, true),
```
Nit -- in other code usually see this as first line of `init/1`

rnewson replied:
agree, we should move this up.

sagelywizard replied:
Done!

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+
+
+init([]) ->
+    ?MODULE = ets:new(?MODULE, [named_table, public, {keypos, #connection.worker}]),
```
Wonder if keeping table protected is a better option, just for safety point of view? Or do we write to it from outside the process, if so then it's ok.

rnewson replied:
we've regretted protected tables in production before. agree that it's quite naughty, but sometimes it's important to be able to write from outside.

nickva replied:
I see. That makes sense.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+
+
+acquire(URL) ->
+    case gen_server:call(?MODULE, {acquire, URL}) of
```
`infinity` here? Saw we do a nice job with async open, but we don't know if would take always < 5 seconds. Or maybe we want the caller to get an error... if so perhaps let them pass in a timeout?

rnewson replied:
yes, this should be infinity (or some configurable value that starts much higher). We might have to create a new connection, and we can't guarantee that will take < 5s.

rnewson replied:
to clarify, that the gen_server opens async only means that other messages can be processed while it's waiting for that reply. The process calling acquire/1 still has to wait for it, and it could taken longer than 5s.

sagelywizard replied:
Do you think we still need this given that the worker creation now happens in the `acquire` call itself?

nickva replied:
Another place it might be possible block is during termination but I see we spawn a process there to stop the client. Besides that just matching on huge ets table and searching throught it might take some time on a full server, but I don't think it should take 5 seconds.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+acquire(URL) ->
+    case gen_server:call(?MODULE, {acquire, URL}) of
+        {ok, Worker} ->
+            link(Worker),
```
If worker dies, this will crash the owner as well. Think that looks good, just wanted to double check. It might be good add a comment in the API of the modules perhaps "acquired worker can crash owner".

sagelywizard replied:
I added this link to minimize necessary changes to the httpc pool. The old httpc pool will crash if an ibrowse worker crashes, so I thought it'd be best to maintain that behavior. Since the link can't happen in the connection sharing process, I put it in the acquire call itself. Can't hurt to add a comment though.

nickva replied:
Makes sense if it is to keep compatibility.

sagelywizard replied:
I added this link to minimize necessary changes to the httpc pool. The old httpc pool would crash if an ibrowse worker crashed, so I thought it'd be best to maintain that behavior. Since the link can't happen in the connection sharing process, I put it in the acquire call itself. Can't hurt to add a comment though.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+create_worker_async(URL) ->
+    spawn_link(fun() ->
+        {ok, Pid} = ibrowse:spawn_link_worker_process(URL),
+        exit({ok, Pid})
```
Nice. Like that it is separated on a new line so exit message shape `{ok, Pid}` is obvious (as opposed to `exit(ibrowse:...)`)

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+                    AsyncPid = create_worker_async(URL),
+                    true = ets:insert_new(
+                        ?MODULE,
+                        #connection{host=Host, port=Port, owner=Pid, worker=AsyncPid}
```
Why insert the record in the connection table if we don't have a connection yet (to keep track of the owner, if they crash while waiting for a acquire reply, so we clean up properly?)

Could we defer inserting this in the table and do it after worker is created, when we reply back?

sagelywizard replied:
That way we can still do that `owner=blah` `match_object` query to drop async workers when the owner dies. So, yes to your question in parens.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+            Response = case Reason of
+                {ok, WorkerPid} ->
+                    [Worker] = ets:lookup(?MODULE, Pid),
+                    ets:insert(?MODULE, Worker#connection{worker=WorkerPid}),
```
Worker is the key, so does it mean we'd have to delete the old tuple which used AsyncPid as key and here insert a new one? (but I might be misremembering how ets insertions work...).  If we didn't put AsyncPid record in ets, we could have grabbed owner from `From`

sagelywizard replied:
Yep, you're right. I also noticed that we need to remove the async procs from the state when the owner crashes as well.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+            end,
+            gen_server:reply(From, Response),
+            {noreply, State#state{waiting=dict:erase(Pid, State#state.waiting)}};
+        error ->
```
Minor nit, if you want pull these separate cases into two functions. As they really do handle two separate things -- death of async opener process and death of a regular worker process.

sagelywizard replied:
Making the separateness of these two clauses explicit sounds good to me, but I generally prefer to express that that via a comment when a bit of code is only used in one place, otherwise it's just spreading logically coupled code through different parts of a module. I don't have a strong opinion though. @rnewson @nickva Thoughts?

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+    } = State,
+    Conns = ets:match_object(?MODULE, #connection{owner=undefined, _='_'}),
+    lists:foreach(fun(Conn) ->
+        ets:delete(?MODULE, Conn#connection.worker),
```
We close all unowned connections every minute. But we don't track how long each connection has been unowned? Just checking if that is what we want. 1 minute is also the default scheduler interval. I think there could be case where we'd close connections which just got unowned, but would have just gotten picked by the new group of rotated processes in the scheduler. But maybe that's too complicated...

We could even just increase the default to be slightly longer than the default scheduling interval. So processes which would rotate back, would get a better chance to pick up their connections again.

sagelywizard replied:
I had the thought of closing connections based on how long they've been left idle, but that would be a fair bit of extra logic and I'm not convinced that it'd solve a real problem, so I went with the simpler solution. I'd propose that we leave this as-is and add more complicated logic if it becomes a problem.

nickva replied:
Agreed. Let's see how it behaves then we can tweak if need it.

rnewson replied:
given tcp windowing, keeping the oldest connections might be a poor strategy anyway.

nickva commented on src/couch_replicator_httpc_pool.erl:
```diff
@@ -31,8 +31,7 @@
     limit,                  % max # of workers allowed
-    free = [],              % free workers (connections)
-    busy = [],              % busy workers (connections)
+    conns = [],
```
Oh nice. Makes things simpler here. At first seems awkward to have a connection pool which also gets items from a connection pool. But it seems this connection pool is used to limit concurrency per replication, which is a useful thing.

sagelywizard replied:
Yeah, my line of thought was that httpc_connection_pool could be setting a connection cap whereas the connection sharing bit would be unbounded. The logic to set a cap on the total number of connections would cause the logic of handling that in the per-replication pooling a fair bit more complicated.

nickva commented on src/couch_replicator_httpc_pool.erl:
```diff
@@ -70,23 +69,15 @@ handle_call(get_worker, From, State) ->
-        [Worker | Free2] ->
-           ok
-        end,
+        {ok, Worker} = couch_replicator_connection:acquire(Url),
```
Nit: Reply can be `error` as well? That's fine, we'd want to crash, but is it worth mentioning in a comment and/or log?

sagelywizard replied:
Sure, I'll add that.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,183 @@
+    {Pid, _Ref} = From,
+    monitor(process, Pid),
+    case ibrowse_lib:parse_url(URL) of
+        #url{host=Host, port=Port} ->
```
Wonder if host includes user and password? Otherwise it would mean sharing connections for  `https://user1:pass1@somecluster.com/db1` and `https://user2:pass2@somecluster.com/db2`?

sagelywizard replied:
Each ibrowse "connection" shouldn't share anything above the TCP level between requests, but we should definitely double check. We could make the connection sharing only happen per-authed-user if we're feeling uncomfortable basing the security of the connection sharing based on the constrait that there's no funny auth-sharing logic in the HTTP client between connections, but that reduces the benefit of connection sharing in quite a few use-cases. Worth discussing on Slack at least. @rnewson @nickva

nickva commented:
@sagelywizard Nice work! Good design. It turned out simpler than I thought.

I only looked over the code so far. Added some comments, some minor nits, some edge cases and such.

rnewson commented:
I can't get this to run. It crashes immediately as waiting is not initialised. If I add waiting=dict:new() in init, things are better but all replications fail.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -97,6 +97,7 @@ handle_call({acquire, URL}, From, State) ->
 handle_cast({relinquish, WorkerPid}, State) ->
     case ets:lookup(?MODULE, WorkerPid) of
         [Worker] ->
+            unlink(Worker),
```
Wonder if owners should do linking / unlinking themselves

sagelywizard replied:
Ah, yeah, this should be unlinked by the owner.

sagelywizard commented:
So, I realized there was a problem last week with the supervision. Also, I dislike that trap_exit, so I was trying to think of a way to get rid of that. I thought I'd make a comment here before writing up a fix.

We want a few constraints:
1) If the owner dies, we want all associated workers to die.
2) If the pooler dies, we want all existing async worker spawners to die.
3) If the pooler dies, we want all existing workers to die.
4) If a worker dies, we want the associated owner to die.

We currently need that trap_exit because we need the pooler to crash the worker spawners and the workers if it dies, but we need to handle `'EXIT'`s when the worker spawners or the workers die.

Right now, there's an awful race condition where if the pooler dies with an async worker spawner response in its inbox, it'll leave the recently created worker orphaned. If we stick all connections under a supervisor with a `let everything die` restart strategy and put that supervisor as well as the pooler under an `one_for_all`, that should take care of `#3`. It'll also make the trap_exit unnecessary, so we can just use monitors and untrapped links.

nickva commented:
@sagelywizard @rnewson 

I dislike `trap_exit` in general but in a few cases, maybe like this it is not too terrible. We are spawning and managing the lifetime of worker processes, we are concerned with catching their 'EXIT's so it makes sense. For example with that flag we get a nice property that workers launched via spawn_link will die if connection pool dies. Without we might have to monitor for parent (pool's) death. 

On constraints. 

1) owner dies => all workers die  : The reason we want all workers to die and not merely returned to the pool is that socket/buffer could end up with unread data from previous owner? If somehow we could be sure a work's state can be cleaned, then we could returned it to the pool (but in reality probably not safe). 

2) pool dies => all worker spawners die: Makes sense. Currently the spawn_link + trap_exit bit will do that, but if we use a sup, we'd have to monitor for it. Another approach is to let them live, they'd finish connecting and exit or send their result to a non-existing pool pip() and die. If this is simpler in code, then I don't see it being too terrible. 

3) pool dies => all workers die: Yap, makes sense.

4) worker dies => owner dies. In general it would be left up to the owner. Maybe they want to monitor.  I know we wanted to limit our changes to the httpc_pool module, but we already made enough changes there that adding a link and unlink on acquire and relinquish might not be too bad, especially if it means one less link and unlink in the new pool process.

Is this a special case to consider as well? "owner dies => all worker spawners die" ? Or was it part of 1). In the code it might be special depending how it is done. If it is simpler, we could postulate that we should just let the worker finish opening than it will be cleaned up if it is not used.  I think this is related to the race condition you mentioned.

Looking back worker spawners made things a bit more complicated because they introduced more possible events and states that we have to consider into this system. Wondering if there is a way to make it simpler. Maybe make the owner process spawn the worker so they are blocked spawning, then tell pool, "Btw I am owner so and so and I spawned and am using this worker. When I die, you can have this worker for other owners to use"?

rnewson commented:
fair points. if trap_exit/link is safer, simpler, better, let's do that.

sagelywizard commented:
Good idea about establishing the connection outside of the pooler proc! I think that makes things simpler.

sagelywizard committed:
Start couch_replicator app when testing httpc pool

sagelywizard committed:
Add tests for replication connection sharing

sagelywizard committed:
Add basic instrumentation to couch_replicator_connection

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,179 @@
+    couch_stats:increment_counter([couch_replicator, connection, owner_crashes]),
+    Workers = ets:match_object(?MODULE, #connection{owner=Pid, _='_'}),
+    lists:foreach(fun(Worker) ->
+        ets:insert(?MODULE, Worker#connection{owner=undefined})
```
I am a bit worried about this. Owner could crash during any random state of reading or writinging the TCP data stream. Do we know for sure that the worker we put back in, can be reused properly by next owner in the state it is in (it might, I am not 100% sure about it though).

We also link workers automatically to owners, there would be a race condition between the owner dying and worker dying as well and getting an EXIT.

In light of automatically linking to owners, maybe monitoring owner for crashes is superfluous (I guess owners could choose to unlink from the owner for whatever reason, then this is fine).

sagelywizard replied:
It'd get removed from the table once the `EXIT` message from the linked workers reached the pooling proc. But I think you're right that monitoring is no longer necessary since we're spawning the ibrowse proc in the `acquire` call.

sagelywizard replied:
Hmm. After thinking about it for a few minutes, I think that there'd still be a race condition if we don't monitor the owner. It could be the case that the owner crashes after acquiring a process, but before linking (e.g. if the reply was sent but the owner timed out). That'd leave the worker allocated to a dead owner.

sagelywizard replied:
It's not a problem when the worker is being created, but it's still a problem if the worker is being reallocated.

sagelywizard replied:
But you're right that we don't want to be reinserting procs into the pool here.

nickva replied:
Thinking about it some more. Monitoring is necessary because owner could exit `normal` and int that case workers won't die (they are linked to owner and if they don't trap exit then they won't notice owner died).

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,179 @@
+    lists:foldl(fun(Conn) ->
+        couch_stats:increment_counter([couch_replicator, connection, closes]),
+        ets:delete(?MODULE, Conn#connection.worker),
+        spawn(fun() -> ibrowse_http_client:stop(Conn) end)
```
Wonder if we need to unlink here. Guessing we call stop in a separate process because we don't want this to block (it could up to 5 seconds).  I am not sure if we reach 5 seconds and ibrowse_http_client will kill the socket, it won't result in another `EXIT` message and logging noise. (Not terrible it will show up as a message from an unknown worker it seems).

For reference, code where ibrowse_http_client kills the socket:

https://github.com/cmullaparthi/ibrowse/blob/master/src/ibrowse_http_client.erl#L105

sagelywizard replied:
Don't need to unlink, but I think it'd be fair.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,179 @@
+    {ok, #state{close_interval=Interval, timer=Timer}}.
+
+
+acquire(URL) ->
```
Minor nit. If we auto-link to the owner, wonder if this should be renamed to `acquire_link(URL)`. Instead of a comment it might be better "explain" that linking will happen via the function name.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,179 @@
+
+
+handle_cast({relinquish, WorkerPid}, State) ->
+    couch_stats:increment_counter([couch_replicator, connection, relinquishs]),
```
Spelling

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,179 @@
+    {noreply, State};
+
+% worker crashed
+handle_info({'EXIT', Pid, Reason}, State) ->
```
Just curious,  would a worker ever exit normally? Hopefully not during regular operation as we plan on recycling them. But wondering if that is possible in some edge cases (owner calls socket close? or owner auto-closes socket on some error...)

sagelywizard replied:
Don't think so, now that I pushed the unlink.

nickva commented on test/couch_replicator_connection_tests.erl:
```diff
@@ -0,0 +1,93 @@
+        couch_replicator_connection ! close_idle_connections,
+        ?assert(ets:member(couch_replicator_connection, Pid)),
+        % block until idle connections have closed
+        sys:get_status(couch_replicator_connection),
```
Ah, nice trick!

nickva commented on test/couch_replicator_connection_tests.erl:
```diff
@@ -0,0 +1,93 @@
+        spawn(fun() ->
+            Self ! couch_replicator_connection:acquire(URL)
+        end),
+        receive
```
Would there be a race condition here? Should we wait for the death of first owner (spawn_monitor then receive 'DOWN' perhaps) to die first.

Then maybe do a variation of the first owner crashing as well.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,184 @@
+
+
+handle_config_change("replicator", "connection_close_interval", V, _, Pid) ->
+    ok = gen_server:cast(Pid, {connection_close_interval, list_to_integer(V)}),
```
I think we forgot to handle this in gen_server.

sagelywizard replied:
Addressed.

nickva commented:
One more -- metrics not defined. Can add them in priv folder. It is not fatal but spams logs with 

`[notice] unknown metric: [couch_replicator,connection,worker_crashes]`

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,189 @@
+
+-export([handle_config_change/5, handle_config_terminate/3]).
+
+-define(DEFAULT_CLOSE_INTERVAL, 60000).
```
Maybe make this slightly bigger than the default scheduling interval, to give connections which rotate time to pick up their old workers (if they are the only ones using it), and avoid running it almost in sync with scheduler as well (both will spawn and start their 1 minute cycles almost in the same phase).

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,189 @@
+    {noreply, State};
+
+% worker crashed
+handle_info({'EXIT', Pid, Reason}, State) ->
```
What about workers that die with normal exit reason? Maybe not bump the `worker_crashes` counter and not log (if those are expected).

I seem to get quite a few of those during testing:

`[info] Replication connection to: "localhost":15984 died with reason normal`

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,189 @@
+delete_worker(Worker) ->
+    ets:delete(?MODULE, Worker#connection.worker),
+    spawn(fun() -> ibrowse_http_client:stop(Worker#connection.worker) end),
+    unlink(Worker#connection.worker),
```
Minor nit, maybe unlink then spawn? I might avoid generating an extra 'EXIT' message perhaps (that saves an extra stats bump and an extra lookup in ets).

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,189 @@
+
+
+handle_cast({relinquish, WorkerPid}, State) ->
+    couch_stats:increment_counter([couch_replicator, connection, relinquishes]),
```
Forgot to demonitor owner. Unlike links monitors will stack, so same owner acquiring worker for same host:port will keep stacking on monitors. We'd need to be a bit smart about it (maybe not monitor again if already monitoring in acquire handle_call...)

sagelywizard replied:
Addressed.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -171,6 +175,33 @@ delete_worker(Worker) ->
+    NewMonitors = case dict:find(Owner, State#state.monitors) of
+        error ->
+            MRef = monitor(process, Owner),
+            dict:store(Owner, {1, MRef}, State#state.monitors);
```
This works. But as a minor nit, was wondering if monitoring can be made simpler by associating a monitor with a connection by Ref. Perhaps, that would avoid, having a to maintain a dict and a ref-count because we refer to owners by pid only.

For example:
- On a acquire/create, before returning we monitor and save ref in ets. Then return.  If owner exits we get the ref from message,  find the worker which matches that ref, delete him.
- On relinquish, we make sure to demonitor (with a [flush] options perhaps) the ref in ets, and set it to undefined.
- If worker exits or crashes, demonitor the ref we find it its ets entry.

sagelywizard replied:
I don't understand how that'd work. Like, if you have an owner acquire workers A, B, and C (in that order) but then relinquished them in order A, B, C, it'd demonitor after it relinquished A, right? Leaving the owner unmonitored while B and C were still owned, right?

nickva replied:
Monitors should stack. So when it relinquishes A it would demonitor RefA. But will still be monitored for B, C by RefB and RefC. The nice thing is monitor refs match 1-to-1 with workers and perhaps we won't have to keep another data structure to map 1-owner-to-many workers by Pid.

I tried it as a quick example:

``` erlang
> P1=spawn(fun() -> receive ok -> ok end end).
> R1=erlang:monitor(process, P1).
> R2=erlang:monitor(process, P1).
> erlang:demonitor(R1).
> exit(P1, kill).
> flush().
Shell got {'DOWN',#Ref<0.0.3.69>,process,<0.48.0>,killed}
ok

> R2.
#Ref<0.0.3.69>
```

So we monitored and demonitored P1 using R1 monitor (demonitoring there is like relinquishing), but R2 was still left and would correctly detect death of P1.

sagelywizard replied:
Oh, I see what you're saying. That is conceptually simpler. I'll make that change.

nickva commented:
Tested with dyno/rep. Data is replicated properly. See connections established in the logs. Monitors are cleaned up. Great work!

Made a note about possibly simplifying reference handling, but I might have also skipped some case or race condition.

Another thing I noticed is  workers exit with reason normal in many cases. So wondering if they are staying alive enough to be meaningfully re-used by others.  I the connection_close_interval to a very large value. Started a few replications with rep. When they stopped replication data, they settled on just a few change feeds. But all the other connections (used for checkpoints, fetching db info) which I expected to to be left in the table sitting un-owned simply exited.

Stats show that re-use did happen (saw `acquires` bumped a large enough times). We can investigate this more tomorrow.

EDIT:

After tracing ibrowse, it turns out when the request is made, we set a connection_timeout from the options which are passed in, normally 120sec. However after a request replies back to caller, the state gets reset and another timeout is set this is for 10sec. So after 10 seconds the worker will get killed. If we manipulate that timeout, we could get a much better pool reuse. It also means our automatic idle connection cleanup might be redundant. (But I vote to keep it). 

Also let's make the tweak how to increase the 10sec timeout a separate PR.  Besides not getting to reuse as many connection, the code seemed to work fine.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,220 @@
+% owner crashed
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
+    couch_stats:increment_counter([couch_replicator, connection, owner_crashes]),
+    Workers = ets:match_object(?MODULE, #connection{owner=Pid, _='_'}),
```
match_delete might work here as well

nickva commented:
@sagelywizard looks much, simpler! eunit tests pass

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -95,21 +95,25 @@ handle_call({create, URL, Worker}, From, State) ->
         [Worker] ->
-            ets:insert(?MODULE, Worker#connection{owner=undefined});
+            case Worker#connection.mref of
+                MRef when is_reference(MRef) -> demonitor(MRef);
```
Maybe add `[flush]`, it would clean the message queue from DOWN messages in case owner died after casting that. Mostly for cosmetic reasons

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -131,16 +134,16 @@ handle_info({'EXIT', Pid, Reason}, State) ->
-                _Pid -> decr_monitor(Owner, State)
-            end
+            case Worker#connection.mref of
+                MRef when is_reference(MRef) -> demonitor(MRef);
```
`[flush]` as well?

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -120,9 +124,8 @@ handle_cast({connection_close_interval, V}, State) ->
 % owner crashed
 handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
```
Wonder how this interacts with one-monitor-per-worker. Technically we could just capture Ref and just clear the worker (if found) which matches the Ref.

If we want to be efficient and kill all workers of a particular owner (by pid), it might make sense to also demonitor + [flush] others, just for proper cleanup.

sagelywizard replied:
All workers will be linked to the owner and will be killed. I don't think it really matters if we kill by pid or ref here.

nickva replied:
But not if owner exits with reason `normal` and somehow never relinquishes the workers. Links will not fire then and we'd have to rely on monitors.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,196 @@
+% owner crashed
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
+    couch_stats:increment_counter([couch_replicator, connection, owner_crashes]),
+    ets:match_delete(?MODULE, #connection{owner=Pid, _='_'}),
```
Let's clean up properly here. It is not much harder, if we just match the mref to Ref. 

In fact there is  no need to keep owner `Pid` around at all. That makes it even cleaner. We refer to owners by Ref only. To workers by Pid.

sagelywizard replied:
Good point. Owner tracking is unnecessary now. Removed.

nickva commented:
Testing with a simple replication job, I see failures -- connection pool crashes taking down everything below (well, to the right of) it.

```
In [1]: rep.replicate_1_to_n_and_wait_till_equal(1)
creating rdyno_src_0000001
 > created  1 dbs with prefix rdyno_src_
creating rdyno_tgt_0000001
 > created  1 dbs with prefix rdyno_tgt_
updating documents
 > _replicator rdyno_0000001_0000001 : 1-93e8a27a73420120da7f949b553cfc6d
>>> update cycle 0  <<<
 > waiting for target  1
 > function _wait_till_dbs_equal succeded after 0.055 +/- 1.0  sec.
 > waiting to propagate changes from  1 to 1  : 0.071 sec.
```

Logs shows _db_changes keeps restarting periodically.

nickva commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,196 @@
+                MRef when is_reference(MRef) -> demonitor(MRef, [flush]);
+                undefined -> ok
+            end,
+            ets:delete(?MODULE, Pid)
```
`ets:delete` returns true. So we make `NewState = true`  -- this kills the ~~crab~~ connection pool and everything after it. 

Deduced from dbg tracing log:

```
{trace,<4641.32113.2>,call,
       {couch_replicator_connection,handle_info,
                                    [close_idle_connections,true]}}
```

We probably don't even need to return NewState here at all, right? We don't modify it.

sagelywizard replied:
Yep! Good catch.

sagelywizard replied:
Don't want to kill the crab. :D

nickva commented:
+1  Tested this. It seems to work. 

Applying this diff, fixes the problem of connection dying too early:

```
diff --git a/src/couch_replicator_connection.erl b/src/couch_replicator_connection.erl
index 7bb68b9..08140f4 100644
--- a/src/couch_replicator_connection.erl
+++ b/src/couch_replicator_connection.erl
@@ -50,6 +50,7 @@ init([]) ->
     ok = config:listen_for_changes(?MODULE, self()),
     Interval = config:get_integer("replicator", "connection_close_interval", ?DEFAULT_CLOSE_INTERVAL),
     {ok, Timer} = timer:send_after(Interval, close_idle_connections),
+    ibrowse:add_config([{inactivity_timeout, Interval}]),
     {ok, #state{close_interval=Interval, timer=Timer}}.


@@ -116,6 +117,7 @@ handle_cast({relinquish, WorkerPid}, State) ->

 handle_cast({connection_close_interval, V}, State) ->
     {ok, cancel} = timer:cancel(State#state.timer),
+    ibrowse:add_config([{inactivity_timeout, V}]),
     {ok, NewTimer} = timer:send_after(V, close_idle_connections),
     {noreply, State#state{close_interval=V, timer=NewTimer}}.
```

Can discuss this a bit more, perhaps, or postpone it.

rnewson commented on src/couch_replicator_connection.erl:
```diff
@@ -0,0 +1,195 @@
+            {ok, Worker};
+        {error, all_allocated} ->
+            {ok, Pid} = ibrowse:spawn_link_worker_process(URL),
+            ok = gen_server:call(?MODULE, {create, URL, Pid}),
```
if creating a new connection takes more than 5 seconds this crashes.

rnewson commented:
+1 after squerge

sagelywizard committed:
Add supervisor entry for connection sharing process

sagelywizard committed:
Make connection sharer configurable

sagelywizard committed:
Make httpc pool use connection sharing pool

sagelywizard committed:
Share connections between replications

This commit adds functionality to share connections between
replications. This is to solve two problems:

- Prior to this commit, each replication would create a pool of
  connections and hold onto those connections as long as the replication
  existed. This was wasteful and cause CouchDB to use many unnecessary
  connections.
- When the pool was being terminated, the pool would block while the
  socket was closed. This would cause the entire replication scheduler
  to block. By reusing connections, connections are never closed by
  clients. They are only ever relinquished. This operation is always
  fast.

This commit adds an intermediary process which tracks which connection
processes are being used by which client. It monitors clients and
connections. If a client or connection crashes, the paired
client/connection will be terminated.

A client can gracefully relinquish ownership of a connection. If that
happens, the connection will be shared with another client. If the
connection remains idle for too long, it will be closed.

sagelywizard commented:
That's the squerge.
